### ISSUE TITLE
Error when using web3signer with a YubiHSM2

### ISSUE THREAD

**Issue by pincente** [Consensys/web3signer#916]
When attempting to use web3signer with a YubiHSM2 with the latest docker image, a Java exception is noted in the logs:

`tk-bs-main-web3signer-1  | 2023-09-18 20:40:48.114+00:00 | main | ERROR | Runner | Error loading signers
tk-bs-main-web3signer-1  | java.util.concurrent.ExecutionException: java.lang.IllegalAccessError: class iaik.pkcs.pkcs11.Module (in unnamed module @0x7690781) cannot access class sun.security.pkcs11.wrapper.CK_C_INITIALIZE_ARGS (in module jdk.crypto.cryptoki) because module jdk.crypto.cryptoki does not export sun.security.pkcs11.wrapper to unnamed module @0x7690781`

However, when the docker image is switched to the 'latest-jdk11' tag, the error is no longer present, and communication with the HSM is as expected:

`tk-bs-main-web3signer-1  | 2023-09-19 01:12:42.159+00:00 | pool-2-thread-1 | INFO  | SignerLoader | Loading signer configuration metadata files from /var/lib/web3signer/keys
tk-bs-main-web3signer-1  | 2023-09-19 01:12:42.163+00:00 | pool-2-thread-1 | INFO  | SignerLoader | Signer configuration metadata files read in memory 1 in 00:00:00.002
tk-bs-main-web3signer-1  | 2023-09-19 01:12:42.196+00:00 | pool-2-thread-1 | INFO  | SignerLoader | Converting signing metadata to Artifact Signer using parallel streams ...
tk-bs-main-web3signer-1  | [P11 - INF 21:12:42.205039] yubihsm_pkcs11.c:266 (C_Initialize): Found 1 configured connector(s)
tk-bs-main-web3signer-1  | [LIB - INF 21:12:42.205051] yubihsm.c:4430 (yh_init_connector): Loading http backend
tk-bs-main-web3signer-1  | [LIB - INF 21:12:42.208460] yubihsm_curl.c:87 (backend_create): backend_create
tk-bs-main-web3signer-1  | [LIB - INF 21:12:42.208475] yubihsm_curl.c:92 (backend_connect): backend_connect
...
`

**Comment by pincente** [comment]
Additional note, in further testing using the 'latest-jdk11' tagged docker image, we were attempting to read a single key from the HSM as a test. As noted earlier, the communication with the HSM seems to be functioning as expected, however there is an error:

`tk-bs-test-hsm-web3signer-1  | [P11 - INF 17:06:24.287445] util_pkcs11.c:5204 (populate_template): Retrieving attribute value, length is 66
tk-bs-test-hsm-web3signer-1  | [P11 - INF 17:06:24.287451] util_pkcs11.c:5224 (populate_template): Attribute/length successfully returned with length 66
tk-bs-test-hsm-web3signer-1  | 2023-10-02 20:06:24.288+00:00 | ForkJoinPool-1-worker-3 | ERROR | SignerLoader | Error converting signing metadata to Artifact Signer: Unexpected error while converting SigningMetadata to ArtifactSigner
` 

In the above case, we've stored a sample key (taken from https://docs.web3signer.consensys.net/reference/key-config-file-params) on the HSM and created a w3s config file with both an entry for the key stored on the HSM, as well as an entry for another sample key in raw unencrypted format. The latter is read correctly:

`tk-bs-test-hsm-web3signer-1  | 2023-10-02 20:06:24.438+00:00 | pool-2-thread-1 | INFO  | RegisteredValidators | Validators registered successfully in database:1
tk-bs-test-hsm-web3signer-1  | 2023-10-02 20:06:24.440+00:00 | pool-2-thread-1 | INFO  | DefaultArtifactSignerProvider | Total signers (keys) currently loaded in memory: 1
`

Not sure if this is an issue with how the key is stored - there aren't many options to store an opaque object on the HSM - or an issue with the older image used for testing. Happy to assist with further testing if it helps.

**Comment by usmansaleem** [comment]
Hi Pincente,

Following are my notes while recreating this issue. The export fix will be coming in via #922. However, you should be able to manually update JAVA_OPTS environment variable to add the export command (mentioned in the following notes). I am using `yhusb://`, however, `http://connector:12345` should work without problems as well. Let me know if this helps out.

- Install YubiHSM tools https://developers.yubico.com/YubiHSM2/Releases/
- Set up a new key (id=3, password=password3) for opaque data management, using the YubiHSM default credentials (authkey=1, password=password)
```shell
yubihsm-shell --connector=yhusb:// --authkey=1 --password=password \
--action=put-authentication-key --new-password=password3 --object-id=3 --domains=1,2,3 \
--capabilities=get-opaque,put-opaque,delete-opaque,export-wrapped,get-pseudo-random,put-wrap-key,import-wrapped \
--delegated=exportable-under-wrap,export-wrapped,import-wrapped

```

The output would look like:
```shell
Created session 0
Stored Authentication key 0x0003
```

* Put opaque data using auth key 3 with the opaque object id is `15`
```shell
yubihsm-shell --connector=yhusb:// --authkey=3 --password=password3 \
--action=put-opaque \
--object-id=15 --label="bls_key_1" --domains=1,2,3 \
--algorithm=opaque-data --capabilities=none \
--informat=hex --in=5e8d5667ce78982a07242739ab03dc63c91e830c80a5b6adca777e3f216a405d
```
The output should look like:
```shell
Session keepalive set up to run every 15 seconds
Created session 0
Stored 32 bytes to Opaque object 0x000f
```

* Use authkey 1 to view the inserted data
```shell
yubihsm-shell --connector=yhusb:// --authkey=1 --password=password --action=get-opaque --object-id=15 --outformat=hex
```
The output would look like:
```shell
Session keepalive set up to run every 15 seconds
Created session 0
5e8d5667ce78982a07242739ab03dc63c91e830c80a5b6adca777e3f216a405d
```

* Identify the location of `yubihsm_pkcs11` dynamic library. For example, on Mac, it is installed at `/usr/local/lib/pkcs11/yubihsm_pkcs11.dylib` using the latest YubiHSM release. 
  Due to a bug in how the yubihsm libraries are generated, this file needs to be copied in its parent folder i.e. `/usr/local/lib`
```bash
sudo cp /usr/local/lib/pkcs11/yubihsm_pkcs11.dylib /usr/local/lib/
```

* In addition, a soft link for `libyubihsm_usb.2.dylib` needs to be created in current directory (from where web3signer is running) i.e.
```bash
ln -s /usr/local/lib/libyubihsm_usb.2.dylib ./libyubihsm_usb.2.dylib
```
 The above changes might not be required on Linux (not tested!)
 
* Create yubihsm configuration file as described at https://docs.web3signer.consensys.net/reference/key-config-file-params#yubihsm-2 under `keys` sub-folder. Take a note of pkcs11 module path. We are using the copied file path i.e. `/usr/local/lib` to bypass Mac issue. Note: The docs needs to be updated, it is referring to incorrect fields in the file. 
```bash
mkdir ./keys

cat <<EOF > ./keys/yubihsm_key1.yaml
type: yubihsm  
connectorUrl: yhusb://  
authId: 3  
password: password3  
opaqueDataId: 15  
pkcs11ModulePath: /usr/local/lib/yubihsm_pkcs11.dylib
additionalInitConfig: debug libdebug timeout=5

EOF
```


* Web3Siger config file `config.yaml`
```bash
cat <<EOF > ./config.yaml
http-listen-host: "0.0.0.0"  
http-listen-port: 9000  
http-host-allowlist: "*"  
  
metrics-enabled: true  
metrics-host: "0.0.0.0"  
metrics-port: 9001  
metrics-host-allowlist: "*"

key-store-path: "./keys/"
  
# eth2 subcommand options  
eth2.network: "minimal"  

eth2.slashing-protection-enabled: false  

EOF
```
* Export JAVA_OPTS with required export command for Web3Signer release without the fix. This will be added in launch scripts in next release of web3signer.
```bash
export JAVA_OPTS="--add-exports jdk.crypto.cryptoki/sun.security.pkcs11.wrapper=ALL-UNNAMED"
```

```
./web3signer --config-file config.yaml eth2
```
The output should look like:
```
2023-10-04 15:30:27.761+10:00 | pool-2-thread-1 | INFO  | SignerLoader | Converting signing metadata to Artifact Signer using parallel streams ...
[P11 - INF 15:30:27.769485] yubihsm_pkcs11.c:266 (C_Initialize): Found 1 configured connector(s)
[LIB - INF 15:30:27.769505] yubihsm.c:4426 (yh_init_connector): Loading usb backend
[LIB - INF 15:30:27.770065] yubihsm_libusb.c:61 (backend_create): backend_create
[LIB - INF 15:30:27.772352] yubihsm_usb.c:48 (backend_connect): backend_connect
[LIB - INF 15:30:27.772364] lib_util.c:178 (parse_usb_url): USB url parsed with serial decimal 0.
...
2023-10-04 15:30:28.176+10:00 | pool-2-thread-1 | INFO  | DefaultArtifactSignerProvider | Total signers (keys) currently loaded in memory: 1
```

**Comment by usmansaleem** [comment]
Fixed via #922. Also, the "hex" private key that needs to be stored as opaque data in YubiHSM should not start with "0x" prefix. Please reopen this issue or create a new one if you have further observations.

**Comment by pincente** [comment]
Thank you @usmansaleem, tests using your approach were successful, and additional tests with our data were also successful.
Some additional notes that linux (or other) users may find helpful:

- The default path where the pkcs11 module is installed via the SDK works in the web3signer config file, ie `/usr/lib/x86_64-linux-gnu/pkcs11/yubihsm_pkcs11.so`
- The soft link for the module is not required
- As noted above, the web3signer config file example in the docs has some issues - specifically `type: yubihsm2` should be `type: yubihsm`, `authKey:` should be `authId:`, and `opaqueObjId:` should be `opaqueDataId:`
- The http connector works as expected

Cheers
